# sc-eQTL
This repository contains the script to parse single-cell RNA-sequencing (scRNA-seq) data and perform downstream analyses like variance partitioning and eQTL mapping from the manuscript [*Refining the resolution of the yeast genotype-phenotype map using single-cell RNA-sequencing*](BioRxiv link). The supplementary files "Matrix_gene_expression_barcodes_1_to_9000.csv" and "Matrix_gene_expression_barcodes_9001_to_18233.csv" contain the expression profile of the single cells. The gene reference/accession indexes are listed in the file "Table_expressed_genes.csv" and the file "Table_single_cell_barcodes_mapping_to_reference_panel_strains_0_based_index.csv" contains the single cell assignment to the reference panel strains (0-based index). In the latter table, the column **best_match** indicates the **0-based index** of the closest reference panel strains to each single cell while the column **significant_best_match** represents the same information but with missing values (NA) when the relatedness between the single cell and its closest strain is not statistically significant (N'Guessan et al., 2023). 

## Pipeline
### I. Expression count
To obtain the expression profile of each single cell, filter out noisy single cells (empty droplets, low number of reads, etc) and obtain the read mapping that allows to perform the allele count, run the following scripts in the specified order:
1. **submit_cellCOUNT_BY_ref.sh**: In this script file, replace the arguments values strating with *"$"* by the appropriate value described at https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count#cr-count where $BY-reference is the path to the reference transcriptome (e.g. https://github.com/arnaud00013/sc-eQTL/tree/main/BY_reference).
2. **submit_cellCOUNT_RM_ref.sh**: In this script file, replace the arguments values strating with *"$"* by the appropriate value described at https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count#cr-count where $RM-reference is the path to the reference transcriptome (e.g. https://github.com/arnaud00013/sc-eQTL/tree/main/RM_reference).

### II. Creating single-cell imputed and corrected genotypes

